hnn_core/optimization/general_optimization.py

```diff
@@ -262,8 +271,9 @@ def _update_params(initial_params, predicted_params):
 
 
 def _run_opt_bayesian(initial_net, tstop, constraints, set_params, obj_fun,
-                      initial_params, max_iter, target, scale_factor=1.,
-                      smooth_window_len=None):
+                      initial_params, max_iter, scale_factor=1.,
+                      smooth_window_len=None, target=None, f_bands=None,
+                      relative_bandpower=None):
```

Comment:
Per our previous discussion, is there a way to pull all parameters related to the objective function (e.g., `scale_factor` and `smooth_window_len`, `target`, etc.) out of the `Optimizer` class?

Reply:
I don't think we can take these out. For example, to optimize evoked drives, we use the `scale_factor` in `_rmse_evoked` (where we call `simulate_dipole`). If we don't let the user pass it in, then amplitude of the optimized dipole most likely will not match the amplitude of the target dipole. In the case of rhythmic drives, we don't really need to pass in `target` (so it is just an optional argument). Are you thinking we should pass these to the `fit` method instead? Or something completely different?

Reply:
could you pass a "partial function" instead of passing each of the arguments?

https://docs.python.org/3/library/functools.html#functools.partial

Reply:
The issue is that the arguments of the optimizer function/class should be general to any type of data simulated with hnn-core. For instance, `smooth_window_len` and `scale_factor` don't make sense if I want to optimize spike rates in the model. And it would be too chaotic to add similar sorts of preprocessing parameters to the API of this function for every conceivable type of objective function for every type of simulated data (spike timing, spike rates, dipole, LFP, CSD, membrane voltage/current, receptor conductance, etc.).

The selection and preprocessing of simulated data, in my mind, should be siloed within the objective function. My recommendation would be to allow the user to either define `obj_fun` as a callable, or pass in `obj_fun_kwargs` that apply to a given objective function defined with a string match. Advanced users can manipulate data preprocessing parameters by defining a partial function (as @jasmainak suggested), and less advanced users can rely on default parameters you define in the template objective functions or set them in  `obj_fun_kwargs` as a catch-all for parameters that are specific to a given objective function.

Reply:
This will also simplify your `_run_opt` functions and `Optimizer.fit()` methods considerably.

Reply:
Ok this makes sense, I just pushed a version that allows users to pass in `obj_fun_kwargs` and I'm working on allowing for a callable now.

Reply:
The user can now pass a callable for `obj_fun`. I'm testing this in `test_general_optimization.py`, `test_user_obj_fun` where `obj_fun` is set to `maxime_csd` which is just a function that maximizes CSD in a certain time range and for a certain electrode depth range.